docs/markdown/Writing Plugins/tutorials/02-integrate-goal-with-the-repository.md

```diff
@@ -0,0 +1,663 @@
+Using own classes via subclassing will also help with refactoring if you decide to deprecate the target alias in order to rename it. In a more advanced scenario, other plugins may import the `ProjectVersionSourceField` field and use it in their own custom targets, so that `project-version` specific behavior would still apply to those targets as well. 
+
+## Ensuring a version follows a semver convention
+
+With the current implementation, we have simply returned the contents of the file as is. We may want to add some validation, for instance, to check that a version string follows a semver convention. Let's learn how to [bring a 3rd party Python package](https://www.pantsbuild.org/docs/plugins-overview#thirdparty-dependencies), namely, [packaging](https://pypi.org/project/packaging/), into our plugin to do that! 
+
+To start depending on the `packaging` package in our in-repo plugin, we must extend the `pants.toml` file:
+
+```toml
+[GLOBAL]
+plugins = ["packaging==22.0"]
+```
+
+Now, let's raise an exception if it isn't possible to construct an instance of the `Version` class:
+
+```python
+from packaging.version import Version, InvalidVersion
+from project_version.target_types import ProjectVersionTarget, ProjectVersionSourceField
+
+class InvalidProjectVersionString(ValueError):
+    pass
+
+@goal_rule
+async def goal_show_project_version(targets: Targets) -> ProjectVersionGoal:
+    targets = [tgt for tgt in targets if tgt.has_field(ProjectVersionSourceField)]
+    results = await MultiGet(
+        Get(ProjectVersionFileView, ProjectVersionTarget, target) for target in targets
+    )
+    for result in results:
+        try:
+           _ = Version(result.version)
+        except InvalidVersion:
+            raise InvalidProjectVersionString(f"Invalid version string '{result.version}' from '{result.path}'")
+    ...
+```
+
+To test this behavior, let's set a bogus version and see our goal in action!
+
+```
+$ cat myapp/VERSION
+x.y.z
+
+$ ./pants project-version myapp:
+[ERROR] 1 Exception encountered:
+
+    InvalidProjectVersionString: Invalid version string 'x.y.z' from 'myapp/VERSION'
+```
+
+## Exploring caching
+
+When you have run the goal a few times, you may have noticed that sometimes the command takes a few seconds to complete, and sometimes it completes immediately. If that's the case, then you have just seen [Pants caching](https://www.pantsbuild.org/docs/rules-api-tips#fyi-caching-semantics) working! Because we use Pants engine to read the `VERSION` file, it copies it into the cache. Pants knows that when the command is re-run, if there are no changes to the Python source code or the `VERSION` file, there's no need to re-run the code because the result is guaranteed to stay the same. If your plugin uses 3rd party Python packages dependencies, it can be worth checking whether the package has any side effects such as reading from the filesystem since this won't let you take full advantage of the Pants engine's caching mechanism.
```

Comment:
I think this is a perfect example of introducing the term "idempotent" and what it means, as that is the property being described here and it's useful to know it as it is used else where in the documentation.

Reply:
Excellent suggestion, done!